bufferserver/src/main/java/com/datatorrent/bufferserver/internal/LogicalNode.java

```diff
@@ -220,17 +220,17 @@ public void catchUp()
       }
 
       if (iterator.hasNext()) {
-        addedData();
+        addedData(false);
       }
     }
 
     logger.debug("Exiting catch up because caughtup = {}", caughtup);
   }
 
   @Override
-  public boolean addedData()
+  public boolean addedData(boolean checkIfListenerHaveDataToSendOnly)
   {
-    if (isReady()) {
+    if (!checkIfListenerHaveDataToSendOnly && isReady()) {
```

Comment:
It depends on the length of an operation that may be added when subscriber is not ready. The call with the flag set to true is made inside synchronized section.